### PR TITLE
IGNITE-12938

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/verify/ValidateIndexesClosure.java [hunk 1]
```diff
@@ -482,28 +476,28 @@ private IndexIntegrityCheckIssue integrityCheckIndexPartition(CacheGroupContext
...
 
     /**
-     * @param grpCtx Group context.
-     * @param part Local partition.
+     * @param grpCtxWithPart Group context and partition id.
      */
     private Map<PartitionKey, ValidateIndexesPartitionResult> processPartition(
-        CacheGroupContext grpCtx,
-        GridDhtLocalPartition part
+        final T2<CacheGroupContext, GridDhtLocalPartition> grpCtxWithPart
```
Comment (ascherbakoff) [inline modules/indexing/src/main/java/org/apache/ignite/internal/visor/verify/ValidateIndexesClosure.java:495]: Wasn't able to understand the purpose of the change. We have one more object as a result.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTaskV2.java [hunk 2]
```diff
@@ -252,97 +250,68 @@ public VerifyBackupPartitionsJobV2(VisorIdleVerifyTaskArg arg) {
 
         /** {@inheritDoc} */
         @Override public Map<PartitionKeyV2, PartitionHashRecordV2> execute() throws IgniteException {
+            try {
+                ignite.context().cache().context().database().waitForCheckpoint("VerifyBackupPartitions");
+            }
+            catch (IgniteCheckedException e) {
+                throw new IgniteException(
+                    "Failed to wait for checkpoint before executing verify backup partitions task", e);
+            }
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTaskV2.java:259]: yep, due to async cluster wide cp, some pages can already synced to storage while others would be still in mem, thus further code : IdleVerifyUtility#checkPartitionsPageCrcSum would return different results. This additional cp prevent from this.

File: modules/indexing/src/main/java/org/apache/ignite/internal/visor/verify/ValidateIndexesClosure.java [hunk 3]
```diff
@@ -199,46 +204,49 @@ private VisorValidateIndexesJobResult call0() {
...
+     *
+     */
+    private VisorValidateIndexesJobResult call0() {
+        Set<Integer> grpIds = collectGroupIds();
+
+        /** Update counters per partition per group. */
+        final Map<Integer, Map<Integer, Long>> partsWithCntrsPerGrp =
+            getUpdateCountersSnapshot(ignite, grpIds);
+
+        IdleVerifyUtility.IdleChecker idleChecker = new IdleVerifyUtility.IdleChecker(ignite, partsWithCntrsPerGrp);
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/indexing/src/main/java/org/apache/ignite/internal/visor/verify/ValidateIndexesClosure.java:242]: We need call this check before all check operations, this near operation :
Map<Integer, IndexIntegrityCheckIssue> integrityCheckResults = integrityCheckIndexesPartitions(grpIds, idleChecker);
need to be wrapped in this check too, isn`t it ?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/IdleVerifyUtility.java [hunk 4]
```diff
@@ -87,28 +95,131 @@ public static void checkPartitionsPageCrcSum(
...
+
+            Map<Integer, Long> partsWithCounters =
+                partsWithCountersPerGrp.computeIfAbsent(grpId, k -> new HashMap<>());
+
+            for (GridDhtLocalPartition part : top.currentLocalPartitions()) {
+                if (part.state() != GridDhtPartitionState.OWNING)
+                    continue;
+
+                if (!part.dataStore().partUpdateCounter().sequential())
+                    throw new GridNotIdleException("Cluster not idle. Probably due to continues rebalance");
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/IdleVerifyUtility.java:132]: Due to lost updates (should not happen under normal circumstances) counter may not be sequential even if grid is idle. Better to compare whole counter before and after check.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTask.java [hunk 5]
```diff
@@ -330,9 +332,9 @@ else if (e.getCause() instanceof IgniteException)
                 long updateCntrAfter = part.updateCounter();
 
                 if (updateCntrBefore != updateCntrAfter) {
-                    throw new IgniteException("Cluster is not idle: update counter of partition [grpId=" +
-                        grpCtx.groupId() + ", partId=" + part.id() + "] changed during hash calculation [before=" +
-                        updateCntrBefore + ", after=" + updateCntrAfter + "]");
+                    throw new GridNotIdleException(GRID_NOT_IDLE_MSG + "[grpName=" + grpCtx.cacheOrGroupName() +
```
Comment (ascherbakoff) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTask.java:335]: The check above will not detect out-of-order updates. We should compare whole counter.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTaskV2.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ascherbakoff) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTaskV2.java:259]: checkPartitionsPageCrcSum is not related in any way to cluster wide cp. 
I think doing a checkpoint not harm anything, let's keep it.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTask.java (hunk 5 above)
Comment (zstan) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/VerifyBackupPartitionsTask.java:335]: VerifyBackupPartitionsTaskV2 now used, VerifyBackupPartitionsTask looks like outdated stuff and need to be removed, i fixed it there:VerifyBackupPartitionsTaskV2.VerifyBackupPartitionsJobV2#calculatePartitionHash

**Pull request closed without merging**